### PR TITLE
fix: prevent user enumeration during registration

### LINKED ISSUE #1582: Registration Flow Allows User Enumeration and Unexpected Deletion of Inactive Accounts
# Problem Statement

The current registration flow automatically deletes inactive accounts that match a submitted username or email address before creating a new account. This behavior can lead to unexpected deletion of legitimate unverified accounts and may allow attackers to infer whether certain usernames or email addresses have previously been registered.

Additionally, concurrent registration attempts involving the same credentials could create race conditions where one user's pending account is unintentionally removed.

# Proposed Solution

Replace automatic deletion of inactive accounts with a safer account recovery or re-verification mechanism. Instead of deleting matching inactive accounts:

* Return a generic message indicating that an account may already exist.
* Allow users to request a new verification email.
* Implement expiration-based cleanup for inactive accounts through scheduled background tasks.
* Ensure registration responses remain consistent regardless of account existence to prevent user enumeration.

# Alternatives Considered

* Retaining the current deletion logic while adding rate limiting.
* Soft-deleting inactive accounts instead of permanently removing them.
* Allowing re-registration only after an inactive account has expired after a defined period.

While these approaches reduce some risks, a dedicated re-verification flow provides a more secure and user-friendly solution.

# Additional Context

Benefits of this change include:

* Preventing accidental deletion of legitimate pending accounts.
* Reducing the risk of username and email enumeration.
* Eliminating potential race conditions during registration.
* Improving overall account security and user experience.

This change would not significantly alter the existing registration workflow while providing stronger security guarantees and more predictable behavior for users.

### PR DESCRIPTION
**Pull request by ssuyashhhh**
## Description

This PR hardens the registration flow by removing automatic deletion of inactive accounts and improving account handling during registration.

Previously, the registration process could delete existing inactive accounts with matching usernames or email addresses before creating a new account. This behavior introduced several concerns:

* Potential user enumeration through registration behavior.
* Accidental deletion of legitimate unverified accounts.
* Race conditions during concurrent registration attempts.
* Poor user experience for users who had not yet completed email verification.

This PR replaces the destructive cleanup logic with a safer workflow that preserves inactive accounts and supports secure re-verification where appropriate.

### Changes Made

* Removed automatic deletion of inactive accounts.
* Improved handling of existing inactive users.
* Added safeguards against username/email enumeration.
* Improved registration flow consistency.
* Added tests covering edge cases and security scenarios.
* Updated relevant documentation and code comments.

## Type of Change

* [x] Bug fix (non-breaking change that fixes an issue)
* [x] Documentation update

## Related Issue

Fixes #1582 
## Checklist

* [x] My code follows the project's style guidelines
* [x] I have performed a self-review of my code
* [x] I have commented my code where necessary
* [x] I have updated the documentation if needed
* [x] My changes generate no new warnings
* [x] I have added tests that prove my fix works
* [x] New and existing tests pass locally

### Security Benefits

* Prevents accidental deletion of inactive accounts.
* Reduces opportunities for username and email enumeration.
* Improves reliability of the verification workflow.
* Eliminates a potential race condition in the registration process.
* Aligns the registration flow more closely with Django security best practices.

### Tested Scenarios

* New account registration.
* Registration with an existing verified account.
* Registration with an existing inactive account.
* Verification and re-verification workflows.
* Concurrent registration attempts using the same credentials.

No breaking changes are expected, and existing authentication functionality remains unaffected.

### CODE REVIEW

**Comment by ssuyashhhh** [comment]
@EDWARD-012 @itsmeesnehaa  do review and merge all checks got passed without merge conflict.

File: game/views.py [hunk 1]
```diff
@@ -588,42 +586,110 @@ def register_view(request):
...
+                # Lock rows to prevent race conditions during concurrent registration attempts
+                existing_user_by_username = User.objects.select_for_update().filter(username__iexact=username).first()
+                existing_user_by_email = User.objects.select_for_update().filter(email__iexact=email).first()
+
+                if not existing_user_by_username and not existing_user_by_email:
+                    # Case A: No conflicts. Save new user atomically.
+                    user = form.save(commit=False)
+                    user.is_active = False
+                    user.save()
+                    user_to_verify = user
```
[the remark this answers is not in the document]
Comment (itsmeesnehaa) [reply game/views.py:690]: make chnages here

File: game/views.py [hunk 2]
```diff
@@ -588,42 +586,110 @@ def register_view(request):
...
+                    else:
+                        # Case C: Email matches an inactive account under a different username.
+                        # Trigger secure re-verification for that inactive account.
+                        user_to_verify = existing_user_by_email
+                elif existing_user_by_username:
+                    if existing_user_by_username.is_active:
+                        conflict_type = 'username_active'
+                    else:
+                        conflict_type = 'username_inactive'
+
```
[the remark this answers is not in the document]
Comment (itsmeesnehaa) [reply game/views.py:720]: resolve this !

File: game/views.py [hunk 3]
```diff
@@ -672,22 +737,74 @@ def register_view(request):
...
+                logger.error(f"Failed to send OTP email: {e}")
                 err_msg = (
                     'Failed to send OTP email. '
                     'Please check your email address and try again.'
                 )
                 messages.error(request, err_msg)
+                request.session.pop('registration_user_id', None)
+                request.session.pop('registration_otp_hash', None)
+                request.session.pop('otp_created_at', None)
+                return redirect('register')
```
[the remark this answers is not in the document]
Comment (itsmeesnehaa) [reply game/views.py]: chnages !

File: game/views.py [hunk 4]
```diff
@@ -672,22 +737,74 @@ def register_view(request):
...
+        else:
+            # Case D: Active account/username conflicts.
+            # Setup dummy session state to prevent timing attacks / enumeration
+            dummy_id = -1
+            dummy_otp = str(secrets.randbelow(900000) + 100000)
+            dummy_hash = hashlib.sha256(f"{dummy_otp}:{settings.SECRET_KEY}".encode()).hexdigest()
+            request.session['registration_user_id'] = dummy_id
+            request.session['registration_otp_hash'] = dummy_hash
+            request.session['otp_created_at'] = time.time()
```
[the remark this answers is not in the document]
Comment (itsmeesnehaa) [reply game/views.py:801]: resolve this issue !

File: game/views.py [hunk 5]
```diff
@@ -672,22 +737,74 @@ def register_view(request):
...
+                            owner_email = existing_user_by_username.email
+                            if owner_email:
+                                subject = 'Checkora Username Alert'
+                                msg_plain = (
+                                    f"Hello! Someone tried to register an account using your username '{username}'.\n\n"
+                                    f"Since you already have an active account, no action is required."
+                                )
+                                send_mail(subject, msg_plain, None, [owner_email], fail_silently=True)
+                except Exception as e:
+                    logger.error(f"Failed to send security alert email: {e}")
```
[the remark this answers is not in the document]
Comment (itsmeesnehaa) [reply game/views.py]: resolve!

File: game/views.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ssuyashhhh) [reply game/views.py]: fix it

File: game/management/commands/cleanup_inactive_users.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+        parser.add_argument(
+            '--retention-hours',
+            type=float,
+            default=24.0,
+            help='Retention period in hours (default: 24.0 hours).'
+        )
+
+    def handle(self, *args, **options):
+        retention_hours = options['retention_hours']
+        threshold = timezone.now() - timedelta(hours=retention_hours)
```
[the remark this answers is not in the document]
Comment (ssuyashhhh) [reply game/management/commands/cleanup_inactive_users.py:20]: fix it

File: game/management/commands/cleanup_inactive_users.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+        User = get_user_model()
+        
+        # Query inactive users created before the threshold
+        inactive_users = User.objects.filter(is_active=False, date_joined__lt=threshold)
+        count = inactive_users.count()
+        
+        if count > 0:
+            self.stdout.write(self.style.WARNING(f"Found {count} inactive users older than {retention_hours} hours. Deleting..."))
+            inactive_users.delete()
+            self.stdout.write(self.style.SUCCESS(f"Successfully deleted {count} inactive user(s)."))
```
[the remark this answers is not in the document]
Comment (ssuyashhhh) [reply game/management/commands/cleanup_inactive_users.py:30]: fix it

File: game/views.py [hunk 8]
```diff
@@ -633,61 +733,107 @@ def register_view(request):
...
+                            owner_email = existing_user_by_username.email
+                            if owner_email:
+                                subject = 'Checkora Username Alert'
+                                msg_plain = (
+                                    f"Hello! Someone tried to register an account using your username '{username}'.\n\n"
+                                    "Since you already have an active account, no action is required."
+                                )
+                                send_mail(subject, msg_plain, None, [owner_email], fail_silently=True)
+                except Exception:
+                    logger.exception("Failed to send security alert email")
```
[the remark this answers is not in the document]
Comment (ssuyashhhh) [reply game/views.py:831]: fix it
